### PR TITLE
fix typelib_cxx_loader override logic

### PR DESCRIPTION
**Pull request by doudou**
Hopefully fixes https://github.com/rock-core/package_set/pull/114 that is hopelessly broken

This time with tests (which found more problems)

### CODE REVIEW

File: test/cxx11_test.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
+require 'test_helpers'
```
Comment (2maz) [inline test/cxx11_test.rb]: I had to use require_relative to get the tests running

File: test/cxx11_test.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                assert_equal 'something', @config.get('typelib_cxx_loader')
+            end
+
+            it "sets the loader to castxml if it is the default on this OS" do
+                Rock.should_receive(:default_loader_is_castxml?).and_return(true)
+                Rock.setup_cxx11_support(@os_package_resolver, @config)
+                assert_equal 'castxml', @config.get('typelib_cxx_loader')
+            end
+
+            it "sets the loader to castxml event if it is not the default on this OS" do
```
Comment (2maz) [inline test/cxx11_test.rb]: event -> even

**Comment by 2maz** [comment]
Tests pass, and looks good on my test bootstrap (xenial/autoproj v2).
However, I would expect gccxml + cxx11 also to work for trusty, but cxx11 enabled is set to false after an 'autoproj reconfigure'.

File: test/cxx11_test.rb (hunk 1 above)
Comment (doudou) [inline test/cxx11_test.rb]: You have to run with `-I. -Itest`, or just "rake test"

**Comment by doudou** [comment]
> However, I would expect gccxml + cxx11 also to work for trusty, but cxx11 enabled is set to false after an 'autoproj reconfigure'.

This was already the previous behavior. There is a check whether clang/castxml is available and if it's not it just does not even ask about C++11.

What do you think should be the behavior ?

File: test/cxx11_test.rb (hunk 1 above)
Comment (doudou) [inline test/cxx11_test.rb]: I've added a README.md to explain how to run the tests

**Comment by 2maz** [comment]
> This was already the previous behavior. There is a check whether clang/castxml is available and if it's not it just does not even ask about C++11.
Yet, it is not clear why checking for cxx11 support actually redirects to only checking for castxml support . 
So, since I might miss the obvious, why is cxx11: false enforced when castxml is not available?

**Comment by doudou** [comment]
> So, since I might miss the obvious, why is cxx11: false enforced when castxml is not available?

I guess that my idea was to avoid C++11 if orogen/typelib had no chance to handle it (which is a bit of a moot point given that C++11 parsing is still not officially available in orogen). I wouldn't mind removing that check though.

**Comment by 2maz** [comment]
Ok, but even then the actual types that orogen need to handle have to be compatible (i.e. don't use C++11 features) while the libraries can still be compiled with cxx11?!
I would suggest to remove that check for now.

**Comment by doudou** [comment]
> Ok, but even then the actual types that orogen need to handle have to be compatible (i.e. don't use C++11 features) while the libraries can still be compiled with cxx11?!

Yes. That was my point when I said

> which is a bit of a moot point given that C++11 parsing is still not officially available in orogen

> I would suggest to remove that check for now.

OK. Will do

**Comment by doudou** [comment]
Updated

File: rock/cxx11.rb [hunk 3]
```diff
@@ -1,8 +1,4 @@
 module Rock
-    def self.has_cxx11_support?(os_package_resolver)
```
Comment (2maz) [inline rock/cxx11.rb]: --> has_castxml_support?

File: rock/cxx11.rb [hunk 4]
```diff
@@ -13,21 +9,15 @@ def self.default_loader_is_castxml?(os_package_resolver)
...
-            config.set 'cxx11', false
-            cxx11_enabled = false
-        end
+        config.declare 'cxx11', 'boolean',
+            default: 'no',
+            doc: "whether C++11 should be enabled for Rock packages"
+        cxx11_enabled = config.get('cxx11')
 
         if !config.has_value_for?('typelib_cxx_loader')
             if cxx11_enabled
```
Comment (2maz) [inline rock/cxx11.rb]: has_castxml_support?

Comment (doudou) [inline rock/cxx11.rb]: Since there's already the castxml-is-default test, I just went for that.

**Comment by 2maz** [comment]
Ok, merged.

**Pull request merged**